### PR TITLE
[ASM] Changes to the collection of usr.id for authenticated clients

### PR DESCRIPTION
**Pull request by robertpi**
## Summary of changes

Change to the automatic instrumentation of common authentication frameworks, to collect details of the authenticated user. This mechanism is only active when ASM is enabled. The full specification is in [this internal RFC](https://docs.google.com/document/d/19VHLdJLVFwRb_JrE87fmlIM5CL5LdOBv4AmLxgdo9qI/edit?pli=1#heading=h.hidcz5iyug2k).

## Reason for change

This will improve our detections of Account Take Overs (ATO).

## Test coverage

Added two new unit test covering basic functionally such creating a hash to anonymize user ids.

Updated several existing integration tests that covered automatic collection of user ids, and add some integration tests to cover configure user id collection via remote config.

### CODE REVIEW

**Comment by andrewlock** [comment]
## Benchmarks Report for appsec :snail:

Benchmarks for #5738 compared to [master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0): 
* 1 benchmarks are **faster**, with geometric mean 1.191
* 1 benchmarks have **fewer** allocations
* 1 benchmarks have **more** allocations


The following thresholds were used for comparing the benchmark speeds:
* Mann–Whitney U test with statistical test for significance of **5%**
* Only results indicating a difference greater than **10%** and **0.3 ns** are considered.

Allocation changes below **0.5%** are ignored.


### Benchmark details


<details><summary><strong>Benchmarks.Trace.Asm.AppSecBodyBenchmark</strong> - Faster :tada: Same allocations :heavy_check_mark:



#### Faster :tada: in #5738

|  Benchmark         | base/diff | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.Asm.AppSecBodyBenchmark.ObjectExtractorSimpleBody&#8209;netcoreapp3.1 | 1.191 | 252.55 | 212.01 | 

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`AllCycleSimpleBody`|net6.0|72.6μs|143ns|553ns|0.0721| 0|0|6&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`AllCycleSimpleBody`|netcoreapp3.1|62.7μs|75.2ns|291ns|0.0938| 0|0|6.94&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`AllCycleSimpleBody`|net472|47.1μs|82.1ns|318ns|1.31| 0|0|8.33&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`AllCycleMoreComplexBody`|net6.0|79.4μs|248ns|961ns|0.127| 0|0|9.5&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`AllCycleMoreComplexBody`|netcoreapp3.1|69.3μs|89.5ns|346ns|0.138| 0|0|10.36&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`AllCycleMoreComplexBody`|net472|54.6μs|46.3ns|179ns|1.87| 0.0271|0|11.84&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`ObjectExtractorSimpleBody`|net6.0|151ns|0.116ns|0.449ns|0.00396| 0|0|280&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`ObjectExtractorSimpleBody`|netcoreapp3.1|252ns|0.212ns|0.794ns|0.0038| 0|0|272&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`ObjectExtractorSimpleBody`|net472|170ns|0.109ns|0.424ns|0.0446| 0|0|281&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`ObjectExtractorMoreComplexBody`|net6.0|3.07μs|1.4ns|5.04ns|0.0536| 0|0|3.78&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`ObjectExtractorMoreComplexBody`|netcoreapp3.1|3.88μs|1.81ns|7.02ns|0.0486| 0|0|3.69&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`ObjectExtractorMoreComplexBody`|net472|3.89μs|2.67ns|10.4ns|0.602| 0.00589|0|3.8&nbsp;KB|
|#5738|`AllCycleSimpleBody`|net6.0|71.6μs|105ns|407ns|0.072| 0|0|6&nbsp;KB|
|#5738|`AllCycleSimpleBody`|netcoreapp3.1|61.4μs|110ns|427ns|0.0924| 0|0|6.94&nbsp;KB|
|#5738|`AllCycleSimpleBody`|net472|47.1μs|36.8ns|138ns|1.32| 0|0|8.33&nbsp;KB|
|#5738|`AllCycleMoreComplexBody`|net6.0|79.3μs|131ns|509ns|0.119| 0|0|9.5&nbsp;KB|
|#5738|`AllCycleMoreComplexBody`|netcoreapp3.1|68.9μs|148ns|574ns|0.137| 0|0|10.36&nbsp;KB|
|#5738|`AllCycleMoreComplexBody`|net472|55.2μs|60.4ns|234ns|1.86| 0.0277|0|11.84&nbsp;KB|
|#5738|`ObjectExtractorSimpleBody`|net6.0|140ns|0.12ns|0.433ns|0.00392| 0|0|280&nbsp;B|
|#5738|`ObjectExtractorSimpleBody`|netcoreapp3.1|212ns|0.227ns|0.881ns|0.00371| 0|0|272&nbsp;B|
|#5738|`ObjectExtractorSimpleBody`|net472|171ns|0.118ns|0.456ns|0.0446| 0|0|281&nbsp;B|
|#5738|`ObjectExtractorMoreComplexBody`|net6.0|2.97μs|2.29ns|8.56ns|0.0536| 0|0|3.78&nbsp;KB|
|#5738|`ObjectExtractorMoreComplexBody`|netcoreapp3.1|4.04μs|2.89ns|11.2ns|0.0506| 0|0|3.69&nbsp;KB|
|#5738|`ObjectExtractorMoreComplexBody`|net472|3.86μs|2.29ns|8.87ns|0.602| 0.00579|0|3.8&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.Asm.AppSecEncoderBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`EncodeArgs`|net6.0|38μs|22.7ns|87.7ns|0.455| 0|0|32.4&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`EncodeArgs`|netcoreapp3.1|54.5μs|28.5ns|110ns|0.436| 0|0|32.4&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`EncodeArgs`|net472|67.4μs|36.4ns|141ns|5.14| 0.0672|0|32.5&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`EncodeLegacyArgs`|net6.0|73.8μs|187ns|724ns|0.0363| 0|0|2.14&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`EncodeLegacyArgs`|netcoreapp3.1|108μs|57.2ns|214ns|0| 0|0|2.14&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`EncodeLegacyArgs`|net472|157μs|95.7ns|371ns|0.313| 0|0|2.15&nbsp;KB|
|#5738|`EncodeArgs`|net6.0|37.3μs|24.2ns|90.5ns|0.453| 0|0|32.4&nbsp;KB|
|#5738|`EncodeArgs`|netcoreapp3.1|54.2μs|24.4ns|91.2ns|0.431| 0|0|32.4&nbsp;KB|
|#5738|`EncodeArgs`|net472|66.7μs|46.7ns|181ns|5.13| 0.0667|0|32.5&nbsp;KB|
|#5738|`EncodeLegacyArgs`|net6.0|76.4μs|26.3ns|98.3ns|0| 0|0|2.14&nbsp;KB|
|#5738|`EncodeLegacyArgs`|netcoreapp3.1|106μs|82.1ns|318ns|0| 0|0|2.14&nbsp;KB|
|#5738|`EncodeLegacyArgs`|net472|158μs|101ns|377ns|0.314| 0|0|2.15&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.Asm.AppSecWafBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`RunWafRealisticBenchmark`|net6.0|184μs|212ns|792ns|0| 0|0|2.42&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`RunWafRealisticBenchmark`|netcoreapp3.1|198μs|345ns|1.33μs|0| 0|0|2.37&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`RunWafRealisticBenchmark`|net472|217μs|128ns|495ns|0.325| 0|0|2.43&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`RunWafRealisticBenchmarkWithAttack`|net6.0|123μs|218ns|846ns|0| 0|0|1.46&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`RunWafRealisticBenchmarkWithAttack`|netcoreapp3.1|129μs|106ns|396ns|0| 0|0|1.45&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`RunWafRealisticBenchmarkWithAttack`|net472|144μs|76.1ns|295ns|0.215| 0|0|1.48&nbsp;KB|
|#5738|`RunWafRealisticBenchmark`|net6.0|185μs|261ns|1.01μs|0| 0|0|2.42&nbsp;KB|
|#5738|`RunWafRealisticBenchmark`|netcoreapp3.1|196μs|127ns|490ns|0| 0|0|2.37&nbsp;KB|
|#5738|`RunWafRealisticBenchmark`|net472|217μs|53.2ns|206ns|0.326| 0|0|2.43&nbsp;KB|
|#5738|`RunWafRealisticBenchmarkWithAttack`|net6.0|121μs|120ns|463ns|0| 0|0|1.46&nbsp;KB|
|#5738|`RunWafRealisticBenchmarkWithAttack`|netcoreapp3.1|130μs|130ns|503ns|0| 0|0|1.45&nbsp;KB|
|#5738|`RunWafRealisticBenchmarkWithAttack`|net472|144μs|49.7ns|193ns|0.215| 0|0|1.48&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.Iast.StringAspectsBenchmark</strong> - Same speed :heavy_check_mark: More allocations :warning:



#### More allocations :warning: in #5738

|  Benchmark         | Base Allocated | Diff Allocated | Change   | Change % |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.Iast.StringAspectsBenchmark.StringConcatAspectBenchmark&#8209;netcoreapp3.1 | 253.43 KB | 265.42 KB | 11.99 KB | 4.73%



#### Fewer allocations :tada: in #5738

|  Benchmark         | Base Allocated | Diff Allocated | Change   | Change % |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.Iast.StringAspectsBenchmark.StringConcatAspectBenchmark&#8209;net6.0 | 264.5 KB | 254.15 KB | -10.35 KB | -3.91%

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`StringConcatBenchmark`|net6.0|58μs|691ns|6.81μs|0| 0|0|43.44&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`StringConcatBenchmark`|netcoreapp3.1|62μs|842ns|8.25μs|0| 0|0|42.64&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`StringConcatBenchmark`|net472|38.5μs|104ns|376ns|0| 0|0|59.27&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`StringConcatAspectBenchmark`|net6.0|306μs|1.59μs|10.3μs|0| 0|0|264.5&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`StringConcatAspectBenchmark`|netcoreapp3.1|342μs|1.83μs|12.8μs|0| 0|0|253.43&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)|`StringConcatAspectBenchmark`|net472|269μs|2.92μs|27.6μs|0| 0|0|278.53&nbsp;KB|
|#5738|`StringConcatBenchmark`|net6.0|60.1μs|696ns|6.92μs|0| 0|0|43.44&nbsp;KB|
|#5738|`StringConcatBenchmark`|netcoreapp3.1|60.3μs|709ns|7.02μs|0| 0|0|42.64&nbsp;KB|
|#5738|`StringConcatBenchmark`|net472|38.6μs|172ns|644ns|0| 0|0|59.07&nbsp;KB|
|#5738|`StringConcatAspectBenchmark`|net6.0|276μs|5.58μs|55.5μs|0| 0|0|254.15&nbsp;KB|
|#5738|`StringConcatAspectBenchmark`|netcoreapp3.1|343μs|1.93μs|13.3μs|0| 0|0|265.42&nbsp;KB|
|#5738|`StringConcatAspectBenchmark`|net472|312μs|7.21μs|69.9μs|0| 0|0|278.53&nbsp;KB|
</details>

**Comment by andrewlock** [comment]
## Throughput/Crank Report:zap:

Throughput results for AspNetCoreSimpleController comparing the following branches/commits:
- [This PR (5738)](https://github.com/DataDog/dd-trace-dotnet/tree/5f1ea98da0d82901bea9adc3c97a40f2b5b0b3eb)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/61603e1639aace563c3b1f950c1c2f083b2d03c0)
- [benchmarks/2.9.0](https://github.com/DataDog/dd-trace-dotnet/tree/89ae3f4ee7d85b824badda5873166707c9e560b4)

Cases where throughput results for the PR are worse than latest master (5% drop or greater), results are shown in **red**.

Note that these results are based on a _single_ point-in-time result for each branch. For full results, see [one](https://ddstaging.datadoghq.com/dashboard/fnz-afi-c2i/apm-net-crank) of the [many](https://ddstaging.datadoghq.com/dashboard/c5g-i7d-kcy/ciapp-apm-net-throughput-tests), [many](https://ddstaging.datadoghq.com/dashboard/uxh-m8j-qhi/ciapp-apm-net-throughput-tests-kevin) dashboards!

```mermaid
gantt
    title Throughput Linux x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (5738) (11.555M)   : 0, 11554814
    master (11.424M)   : 0, 11424238
    benchmarks/2.9.0 (11.598M)   : 0, 11597798

    section Automatic
    This PR (5738) (7.857M)   : 0, 7857034
    master (7.758M)   : 0, 7758399
    benchmarks/2.9.0 (8.205M)   : 0, 8204910

    section Trace stats
    master (8.022M)   : 0, 8021970

    section Manual
    This PR (5738) (10.048M)   : 0, 10048387
    master (9.878M)   : 0, 9877888

    section Manual + Automatic
    This PR (5738) (7.463M)   : 0, 7462568
    master (7.319M)   : 0, 7319136

    section Version Conflict
    master (6.362M)   : 0, 6361671

```
```mermaid
gantt
    title Throughput Linux arm64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (5738) (9.456M)   : 0, 9455917
    master (9.424M)   : 0, 9424194
    benchmarks/2.9.0 (9.629M)   : 0, 9629283

    section Automatic
    This PR (5738) (6.442M)   : 0, 6442445
    master (6.492M)   : 0, 6491874

    section Trace stats
    master (6.890M)   : 0, 6890499

    section Manual
    This PR (5738) (8.403M)   : 0, 8402570
    master (8.159M)   : 0, 8158813

    section Manual + Automatic
    This PR (5738) (6.184M)   : 0, 6184171
    master (6.119M)   : 0, 6119258

    section Version Conflict
    master (5.657M)   : 0, 5656618

```
```mermaid
gantt
    title Throughput Windows x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (5738) (10.107M)   : 0, 10107118
    master (10.199M)   : 0, 10198885
    benchmarks/2.9.0 (10.219M)   : 0, 10219220

    section Automatic
    This PR (5738) (7.212M)   : 0, 7211959
    master (7.214M)   : 0, 7213911
    benchmarks/2.9.0 (7.436M)   : 0, 7435759

    section Trace stats
    master (7.557M)   : 0, 7556605

    section Manual
    This PR (5738) (8.996M)   : 0, 8995805
    master (9.196M)   : 0, 9195549

    section Manual + Automatic
    This PR (5738) (6.912M)   : 0, 6911639
    master (7.067M)   : 0, 7067296

    section Version Conflict
    master (6.248M)   : 0, 6248496

```

File: tracer/src/Datadog.Trace/AppSec/SecuritySettings.cs [hunk 1]
```diff
@@ -55,7 +60,7 @@ public SecuritySettings(IConfigurationSource? source, IConfigurationTelemetry te
             WafTimeoutMicroSeconds = (ulong)config
                                            .WithKeys(ConfigurationKeys.AppSec.WafTimeout)
                                            .GetAs<int>(
-                                                getDefaultValue: () => 100_000, // Default timeout of 100 ms, only extreme conditions should cause timeout
+                                                getDefaultValue: _ => 100_000, // Default timeout of 100 ms, only extreme conditions should cause timeout
```
Comment (daniel-romano-DD) [inline tracer/src/Datadog.Trace/AppSec/SecuritySettings.cs:63]: Why this change? Out of curiosity

File: tracer/test/Datadog.Trace.Security.Unit.Tests/SecurityUserTrackingModeTests.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    // ident            | undefined         | identification
+    // anon             | undefined         | anonymization
+    //
+    // The library doesn't use these values directly, but two booleans calculated from them:
+    //  - IsTrackUserEventsEnabled
+    //  - IsAnonUserTrackingMode
+    // This test suite demonstrates that these two booleans are correctly calculated from the above table.
+    // Local values can't be null, as the default is calculated within the framework for local config
+    // telemetry.
+    // Remote config can be null, and this tell us remote config is not active.
```
Comment (daniel-romano-DD) [inline tracer/test/Datadog.Trace.Security.Unit.Tests/SecurityUserTrackingModeTests.cs:35]: This is super helpful 👍

File: tracer/test/Datadog.Trace.Tests/Telemetry/config_norm_rules.json [hunk 3]
```diff
@@ -77,7 +77,7 @@
   "tracer_instance_count": "trace_instance_count",
   "trace.db.client.split-by-instance": "trace_db_client_split_by_instance",
   "trace.db.client.split-by-instance.type.suffix": "trace_db_client_split_by_instance_type_suffix",
-  "trace.http.client.split-by-domain": "trace_http_client_split_by_domain",
+  "trace.http.client.split-by-domain" : "trace_http_client_split_by_domain",
```
Comment (daniel-romano-DD) [inline tracer/test/Datadog.Trace.Tests/Telemetry/config_norm_rules.json:80]: Why this extra space?

**Review by daniel-romano-DD — APPROVED** [review]
Didn't delve to the bottom of the logic, but coding looks good to me.

File: tracer/src/Datadog.Trace/AppSec/Security.cs [hunk 4]
```diff
@@ -136,12 +136,38 @@ public static Security Instance
...
+            {
+                return true;
+            }
+
+            if (remote is SecuritySettings.UserTrackingDisabled or not null)
+            {
+                return false;
+            }
+
+            return local is SecuritySettings.UserTrackingIdentMode or SecuritySettings.UserTrackingAnonMode;
```
Comment (NachoEchevarria) [inline tracer/src/Datadog.Trace/AppSec/Security.cs:158]: If local is undefined and remote is also undefined, should we return true (identification mode enabled)?

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AspNetCore/UserEvents/AuthenticationHttpContextExtensionsIntegration.cs [hunk 5]
```diff
@@ -54,42 +56,46 @@ internal static object OnAsyncMethodEnd<TTarget>(object returnValue, Exception e
         {
             var claimsPrincipal = state.State as ClaimsPrincipal;
             // https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims
-            var claimsToTestInSafeMode = new[] { ClaimTypes.NameIdentifier, ClaimTypes.Name, "sub" };
-            if (claimsPrincipal?.Claims != null && Security.Instance is { TrackUserEvents: true } security)
+            var claimsToTest = new[] { ClaimTypes.NameIdentifier, ClaimTypes.Name, "sub", ClaimTypes.Email, ClaimTypes.Name, };
```
Comment (NachoEchevarria) [inline tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AspNetCore/UserEvents/AuthenticationHttpContextExtensionsIntegration.cs:59]: A performance improvement could be defining ClaimsToTest as a static hashset:

        private static readonly HashSet<string> ClaimsToTest = new HashSet<string> 
        { 
            ClaimTypes.NameIdentifier, ClaimTypes.Name, "sub", ClaimTypes.Email 
        };

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AspNetCore/UserEvents/AuthenticationHttpContextExtensionsIntegration.cs [hunk 6]
```diff
@@ -54,42 +56,46 @@ internal static object OnAsyncMethodEnd<TTarget>(object returnValue, Exception e
...
+
+                        break;
                     }
                 }
 
-                security.SetTraceSamplingPriority(span);
+                if (foundUserId)
+                {
+                    security.SetTraceSamplingPriority(span);
+                    setTag(Tags.AppSec.EventsUsers.LoginEvent.SuccessTrack, "true");
```
Comment (NachoEchevarria) [inline tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AspNetCore/UserEvents/AuthenticationHttpContextExtensionsIntegration.cs:93]: Is SuccessTrack ever set to false?

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AspNetCore/UserEvents/UserEventsCommon.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+        // spec says take first half of the hash
+        const int bytesToUse = 16;
+
+        if (byteArray.Length >= bytesToUse)
+        {
+            // we want to end up with a string of the form anon_0c76692372ebf01a7da6e9570fb7d0a1
+            // 37 is 5 prefix character plus 32 hexadecimal digits (presenting 16 bytes)
+            // stackalloc to avoid intermediate heap allocation
+            var stringChars = stackalloc char[37];
```
Comment (NachoEchevarria) [inline tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AspNetCore/UserEvents/UserEventsCommon.cs:35]: Nice!

File: tracer/test/snapshots/Security.AspNetCore5AutoUserEvents.SecurityOn.defaultmode.TestUserLoginEvent_eventName=login.auto.success_bodyString=Input.UserName=TestUser&Input.Password=test.verified.txt [hunk 8]
```diff
@@ -24,7 +24,7 @@
       runtime-id: Guid_1,
       span.kind: server,
       usr.id: Guid_2,
-      _dd.appsec.events.users.login.success.auto.mode: safe,
+      _dd.appsec.events.users.login.success.auto.mode: ident,
```
Comment (NachoEchevarria) [inline tracer/test/snapshots/Security.AspNetCore5AutoUserEvents.SecurityOn.defaultmode.TestUserLoginEvent_eventName=login.auto.success_bodyString=Input.UserName=TestUser&Input.Password=test.verified.txt:27]: Shouldn't safe get translated into anon?

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AspNetCore/UserEvents/AuthenticationHttpContextExtensionsIntegration.cs (hunk 6 above)
Comment (robertpi) [reply tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AspNetCore/UserEvents/AuthenticationHttpContextExtensionsIntegration.cs:93]: No, false is absent.

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AspNetCore/UserEvents/AuthenticationHttpContextExtensionsIntegration.cs (hunk 5 above)
Comment (robertpi) [reply tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AspNetCore/UserEvents/AuthenticationHttpContextExtensionsIntegration.cs]: Fixed in 266b7a62b.

File: tracer/src/Datadog.Trace/AppSec/Security.cs (hunk 4 above)
Comment (robertpi) [reply tracer/src/Datadog.Trace/AppSec/Security.cs:159]: Local can never be null, we handle defaults in the setting class (so they will be recorded by telemetry). Added a comment. See 266b7a62b.

File: tracer/test/snapshots/Security.AspNetCore5AutoUserEvents.SecurityOn.defaultmode.TestUserLoginEvent_eventName=login.auto.success_bodyString=Input.UserName=TestUser&Input.Password=test.verified.txt (hunk 8 above)
Comment (robertpi) [reply tracer/test/snapshots/Security.AspNetCore5AutoUserEvents.SecurityOn.defaultmode.TestUserLoginEvent_eventName=login.auto.success_bodyString=Input.UserName=TestUser&Input.Password=test.verified.txt]: This file tests the default mode, the previous default was "safe", it's now "ident".

When "safe" is explicitly set, it should be translated to "anon". We verify that happens in the unit test, there were no explicit integration tests for safe mode, so I decided not add them. I think the unit tests are enough.

**Review by NachoEchevarria — APPROVED** [review]
LGTM!

File: tracer/src/Datadog.Trace/AppSec/SecuritySettings.cs [hunk 9]
```diff
@@ -67,15 +72,40 @@ public SecuritySettings(IConfigurationSource? source, IConfigurationTelemetry te
...
+                        val =>
+                            val.Equals(UserTrackingDisabled, StringComparison.OrdinalIgnoreCase)
+                         || val.Equals(DeprecatedUserTrackingSafeMode, StringComparison.OrdinalIgnoreCase)
+                         || val.Equals(DeprecatedUserTrackingExtendedMode, StringComparison.OrdinalIgnoreCase))
+                   .ToLowerInvariant();
+
+            UserEventsAutoInstrumentationMode = config
+                                               .WithKeys(ConfigurationKeys.AppSec.UserEventsAutoInstrumentationMode)
+                                               .AsString(
+                                                    isPresent => isPresent ? UserTrackingDisabled : GetLegacyUserTracking(),
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/AppSec/SecuritySettings.cs:89]: This is somewhat of a fundamental change to the design of the `ConfigurationBuilder` and I'm not sure we should make it 😬 Also, it's kinda hard to follow, what behaviour are we trying to get here? I'm 90% sure we can do it without changing the builder...

Comment (robertpi) [reply tracer/src/Datadog.Trace/AppSec/SecuritySettings.cs:89]: The config setting name was changed, if the new value is present we should use it, otherwise we should look for the old one.

The change allows us to do this all inside the configuration combinators, we could implement this logic without the changer but it would mean moving some of the defaults outside of the combinators so the changes would be visible to the telemetry.

Comment (andrewlock) [reply tracer/src/Datadog.Trace/AppSec/SecuritySettings.cs:89]: > The config setting name was changed, if the new value is present we should use it, otherwise we should look for the old one.

We have that capability already, by passing the key values to the `WithKeys()` and it's handled correctly by telemetry in that case 🤔 

e.g. see here, where we have two "deprecated" values that we fallback to: https://github.com/DataDog/dd-trace-dotnet/blob/0f912fefdd124405fe162c99458c9103ada53d82/tracer/src/Datadog.Trace/Configuration/TracerSettings.cs#L394

Comment (robertpi) [reply tracer/src/Datadog.Trace/AppSec/SecuritySettings.cs]: As discussed on slack used `.AsStringResult` and rolled back the changes.

**Review by bouwkast — APPROVED** [review]
Focused mainly on the IDM stuff but looked good to me, thanks!
